retrieving/retrieving-and-querying-mongodb-documents-with-go.md

```diff
@@ -0,0 +1,157 @@
+# Quick Start: Retrieving and Querying MongoDB Documents with Go
+
+In a [previous tutorial](https://), I had written about creating Documents to be inserted into MongoDB with the Go programming language. In that tutorial we explored the `Insert` and `InsertMany` functions while making use of `bson.D`, `bson.M`, and similar MongoDB data types.
+
+This time around, we're going to explore reading Documents from MongoDB and creating queries to retrieve Documents based on certain criteria. This will all be done with Golang and the MongoDB Go Driver.
+
+## Tools and Versions for the Tutorial Series
+
+I wanted to take a moment to reiterate the tools and versions that I'm using within this tutorial series:
+
+- Go 1.13
+- Visual Studio Code (VS Code)
+- MongoDB Atlas with an M0 free cluster
+- MongoDB Go Driver 1.1.2
+
+To get the best experience while following this tutorial, try to match the versions as best as possible. However, other versions may still work without issue.
+
+> You can get started with an M0 cluster on [MongoDB Atlas](https://www.mongodb.com/cloud) for free. If sign up using the promotional code NRABOY200, you'll receive premium credit applied to your account.
+
+If you need help connecting to MongoDB Atlas, installing the MongoDB Go Driver, or getting familiar with creating Documents, I encourage you to check out one of the previous tutorials in the series.
+
+## The Previously Created Data
+
+When thinking back to the data that we created in the [previous tutorial](https://), we know that we have a `podcasts` collection with data that looks something like this:
+
+```json
+{
+    "_id": ObjectId("5d9e0173c1305d2a54eb431a"),
+    "title": "The Polyglot Developer Podcast",
+    "author": "Nic Raboy"
+}
+```
+
+We also have an `episodes` collection which has similarly structured data that looks like this:
+
+```json
+{
+    "_id": ObjectId("5d9f4701e9948c0f65c9165d"),
+    "podcast": ObjectId("5d9e0173c1305d2a54eb431a"),
+    "title": "GraphQL for API Development",
+    "description": "Learn about GraphQL development in this episode of the podcast.",
+    "duration": 25
+}
+```
+
+Knowing the fields that exist in our data will be important to us when it comes to crafting queries to return only the data that we need, rather than everything.
+
+## Reading All Documents from a Collection
+
+Reading all data from a collection consists of making the request, then looping through the results. Knowing what fields exist on each of the documents isn't too important, only knowing the collection name itself.
+
+A simple example of this can be done through the following:
+
+```golang
+cursor, err := episodesCollection.Find(ctx, bson.M{})
+if err != nil {
+    log.Fatal(err)
+}
+for cursor.Next(ctx) {
+    var episode bson.D
+    cursor.Decode(&episode)
+    fmt.Println(episode)
+}
+cursor.Close(ctx)
```

Comment:
If using the loop, this `cursor.Close` call should be moved up and called as a `defer`:
```
cursor, err := episodesCollection.Find(ctx, bson.M{})
if err != nil {
    log.Fatal(err)
}
defer cursor.Close(ctx)
```

Reply:
Wouldn't both of these practices accomplish the same thing? I figured in the scenario where the cursor might be set numerous times from a `Find`, it should first be closed between usages. With the `defer` it would only close when the parent function terminates. Right?

Reply:
In the event that an error or panic occurs between the code creates the cursor and the call to `Close`, there will be a connection/session leak. This isn't an issue with your original code, but one of my comments is to check the error from `Decode`. Doing that introduces a new error path.

Reply:
If we catch the errors and make use of `defer`, can the cursor be reset even if it isn't explicitly closed first?

Reply:
I'm not sure off the top of my head. All of this goes away if you use `cursor.All` as that closes the cursor for you. I think that's the best course of action here.